### PR TITLE
feat(rust,python,cli): add SQL support for `UNION [ALL] BY NAME`, add "diagonal_relaxed" strategy for `pl.concat`

### LINKED ISSUE #9370: `UNION` too sensitive with column ordering
### Polars version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Issue description

The `UNION` clause throws an exception when being applied on two `DataFrame`s that have exactly the same column names (and data types) but different ordering. This relates to #8936 .

### Reproducible example

```python
import polars as pl

df1 = pl.DataFrame({"a": [1], "b": ['x']})
df2 = pl.DataFrame({"b": ['y'], "a": [2]})

sql_ctx = pl.SQLContext(register_globals=True, eager_execution=True)

sql_ctx.execute("""
SELECT * FROM df1
UNION
SELECT * FROM df2
""")

# Output
ShapeError: unable to vstack, column names don't match: "a" and "b"
```

### Expected behavior

```shell
shape: (2, 2)
┌─────┬─────┐
│ a   ┆ b   │
│ --- ┆ --- │
│ i64 ┆ str │
╞═════╪═════╡
│ 1   ┆ x   │
│ 2   ┆ y   │
└─────┴─────┘
```

### Installed versions

<details>

```
--------Version info---------
Polars:      0.18.2
Index type:  UInt32
Platform:    macOS-13.3.1-arm64-arm-64bit
Python:      3.9.16 (main, May 16 2023, 20:00:19) 
[Clang 14.0.3 (clang-1403.0.22.14.1)]

----Optional dependencies----
numpy:       1.24.3
pandas:      2.0.2
pyarrow:     10.0.1
connectorx:  <not installed>
deltalake:   <not installed>
fsspec:      <not installed>
matplotlib:  <not installed>
xlsx2csv:    <not installed>
xlsxwriter:  <not installed>
```

</details>

### LINKED ISSUE #9412: Relaxed concatenation for "diagonal"
### Problem description

One and a half years ago i asked if it was possibly to implement diagonal concatenation with auto cast of not matching dtypes of the same columns in different DataFrames. The statement from that time was "Auto-Casting of identical columns with different dtypes shouldn't exist" but diagonal stacking was still implemented which i'm still thankful for.
With https://github.com/pola-rs/polars/pull/9382 now in place (relaxed concatenation including auto-cast) there seems to have been a change of mind so i would like to renew my question if it would now be possible to implement relaxed concatenation option for diagonal concatenation.
The details with explanations can be found in the old issue https://github.com/pola-rs/polars/issues/1665.

### LINKED ISSUE #9891: Auto-sort column order when concat `vertical` and `vertical_relaxed`
### Problem description

As of `0.18.7`, Polars will return an error if columns' order does not align with one another when concat vertically.

```python
import polars as pl

df1 = pl.DataFrame({'a': [1, 2], 'b': [2, 1]})
df2 = pl.DataFrame({'b': [1, 2], 'a': [None, 1]})
```
```python
pl.concat([df1, df2], how='vertical')
# ShapeError: unable to vstack, column names don't match: "a" and "b"
```

```python
pl.concat([df1, df2], how='vertical_relaxed')
# ComputeError: schema names differ: got a, expected b
```

It would be great if polars could sort the columns itself.

### PR DESCRIPTION
**Pull request by alexander-beedie**
# Issues closed

⭐  _"Three for the price of one..."_
Additional ref: https://github.com/pola-rs/polars/issues/8917#issuecomment-1687013214.

Closes #9370.
Closes #9891.
Closes #9412.

# TLDR

* New `pl.concat` strategy "diagonal_relaxed".
* New SQL support, eg: `SELECT * FROM df1 UNION ALL BY NAME SELECT * FROM df2`.
* The "to_supertypes" parameter is set `True` by default for `UNION` ops _iif_ coming from the SQL interface, as this is the expected behaviour in that environment.

# Miscellaneous

While adding diagonal support for the "to_supertypes" param I also standardised the Rust-side function names for consistency/clarity, and updated the diagonal variant signature to match the standard concat, such that they both now take exactly the same `UnionArgs` enum (instead of breaking out the args as standalone params, as was previously done).

* `hor_concat_df` → `concat_df_horizontal`
* `diag_concat_df` → `concat_df_diagonal`
* `diag_concat_lf` → `concat_lf_diagonal`

# Examples

* Setup sample frame data:
  
  ```python
  import polars as pl
  
  df1 = pl.DataFrame({
      "A": [1, 2, 2],
      "B": [5, 4, 4],
  })
  
  df2 = df1.select(
      pl.col("B").cast(pl.UInt16),
      pl.col("A").cast(pl.Int32),
      pl.lit(123).alias("C"),
  )
  ```

* Demonstrate new `pl.concat` strategy:

  ```python
  pl.concat( [df1,df2], how="diagonal" )
  # ShapeError: 
  #  unable to vstack, dtypes for column "A" don't match: `i64` and `i32`

  pl.concat( [df1,df2], how="diagonal_relaxed" )
  # ┌─────┬─────┬──────┐
  # │ A   ┆ B   ┆ C    │
  # │ --- ┆ --- ┆ ---  │
  # │ i64 ┆ i64 ┆ str  │
  # ╞═════╪═════╪══════╡
  # │ 1   ┆ 5   ┆ null │
  # │ 2   ┆ 4   ┆ null │
  # │ 2   ┆ 4   ┆ null │
  # │ 1   ┆ 5   ┆ 0    │
  # │ 2   ┆ 4   ┆ 0    │
  # │ 2   ┆ 4   ┆ 0    │
  # └─────┴─────┴──────┘
  ```

* Demonstrate equivalent/new SQL support:

  ```python
  with pl.SQLContext(
      register_globals = True, 
      eager_execution = True,
  ) as ctx:
  
      ctx.execute(
          "SELECT * FROM df1 UNION ALL BY NAME SELECT * FROM df2"
      )
      # ┌─────┬─────┬──────┐
      # │ A   ┆ B   ┆ C    │
      # │ --- ┆ --- ┆ ---  │
      # │ i64 ┆ i64 ┆ str  │
      # ╞═════╪═════╪══════╡
      # │ 1   ┆ 5   ┆ null │
      # │ 2   ┆ 4   ┆ null │
      # │ 2   ┆ 4   ┆ null │
      # │ 1   ┆ 5   ┆ 0    │
      # │ 2   ┆ 4   ┆ 0    │
      # │ 2   ┆ 4   ┆ 0    │
      # └─────┴─────┴──────┘
  
      ctx.execute(
          "SELECT * FROM df1 UNION BY NAME SELECT * FROM df2"
      )
      # ┌─────┬─────┬──────┐
      # │ A   ┆ B   ┆ C    │
      # │ --- ┆ --- ┆ ---  │
      # │ i64 ┆ i64 ┆ i32  │
      # ╞═════╪═════╪══════╡
      # │ 2   ┆ 4   ┆ null │
      # │ 1   ┆ 5   ┆ null │
      # │ 2   ┆ 4   ┆ 123  │
      # │ 1   ┆ 5   ┆ 123  │
      # └─────┴─────┴──────┘
  ```

---

Bonus: while down this rabbit hole I also made an [upstream PR](https://github.com/sqlparser-rs/sqlparser-rs/pull/997) with `sqlparser-rs` to support `UNION DISTINCT BY NAME`; we don't need to wait on that as the `DISTINCT` keyword is (functionally) optional.

### CODE REVIEW

File: crates/polars-lazy/src/dsl/functions.rs [hunk 1]
```diff
@@ -118,12 +118,11 @@ pub(crate) fn concat_impl<L: AsRef<[LazyFrame]>>(
...
-pub fn diag_concat_lf<L: AsRef<[LazyFrame]>>(
-    lfs: L,
-    rechunk: bool,
-    parallel: bool,
+pub fn concat_lf_diagonal<L: AsRef<[LazyFrame]>>(
+    inputs: L,
+    args: UnionArgs,
 ) -> PolarsResult<LazyFrame> {
-    let lfs = lfs.as_ref().to_vec();
+    let lfs = inputs.as_ref().to_vec();
```
Comment (ritchie46) [inline crates/polars-lazy/src/dsl/functions.rs:125]: I see this wasn't introduced here, but let's fix it now. 

We don't need to allocate a Vec. We can clone the `LazyFrame` upon iteration. This saves the heap allocation of the vec with pointers to the logical plan.

**Review by ritchie46 — CHANGES_REQUESTED** [review]
Nice! Small remark that saves a heap allocations. Other than that, good to go. :+1: 

I think many people like a relaxed diagonal. :)

File: crates/polars-lazy/src/dsl/functions.rs (hunk 1 above)
Comment (alexander-beedie) [reply crates/polars-lazy/src/dsl/functions.rs]: Done ✌️:sunglasses:

**Comment by alexander-beedie** [comment]
> I can make a PR if needed, or perhaps it's simpler for you to add it as it is a 1 word change.

Sounds like a plan; I'll be adding/fixing batched `read_database` mode for Databricks SQL tomorrow (now I have access to a test account), so I can throw this in at the same time :)
